sql: TestQueryCache failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/c59f5347d5424edb90575fb0fd50bad677953752

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestQueryCache PKG=github.com/cockroachdb/cockroach/pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1196029&tab=buildLog

```
I190324 05:56:51.773067 24959 sql/event_log.go:135  [n1,client=127.0.0.1:45486,user=root] Event: "create_table", target: 53, info: {TableName:db1.public.t Statement:CREATE TABLE db1.public.t (a INT8, b INT8) User:root}
I190324 05:56:51.775002 24936 storage/replica_command.go:246  [n1,split,s1,r20/1:/{Table/24-Max}] initiating a split of this range at key /Table/53 [r21] (zone config)
W190324 05:56:51.784399 25031 storage/intentresolver/intent_resolver.go:509  failed to push during intent resolution: failed to push id=1ae9ba8a key=/Local/Range/Table/24/RangeDescriptor rw=true pri=0.02187383 stat=PENDING epo=0 ts=1553407011.775016758,0 orig=1553407011.775016758,0 max=0.000000000,0 wto=false seq=10
I190324 05:56:51.789269 24959 sql/event_log.go:135  [n1,client=127.0.0.1:45486,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:sql.query_cache.enabled Value:true User:root}
I190324 05:56:51.828313 24586 util/stop/stopper.go:546  quiescing; tasks left:
10     [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828375 24586 util/stop/stopper.go:546  quiescing; tasks left:
10     [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828451 24586 util/stop/stopper.go:546  quiescing; tasks left:
9      [async] storage.IntentResolver: cleanup txn records
1      [async] intent_resolver_ir_batcher
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828481 24586 util/stop/stopper.go:546  quiescing; tasks left:
9      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828507 24586 util/stop/stopper.go:546  quiescing; tasks left:
8      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828538 24586 util/stop/stopper.go:546  quiescing; tasks left:
7      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828575 24586 util/stop/stopper.go:546  quiescing; tasks left:
6      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828602 24586 util/stop/stopper.go:546  quiescing; tasks left:
5      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828627 24586 util/stop/stopper.go:546  quiescing; tasks left:
4      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828657 24586 util/stop/stopper.go:546  quiescing; tasks left:
3      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828683 24586 util/stop/stopper.go:546  quiescing; tasks left:
2      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828705 24586 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] storage.IntentResolver: cleanup txn records
1      [async] closedts-rangefeed-subscriber
I190324 05:56:51.828738 24586 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
    --- FAIL: TestQueryCache/statschange (0.24s)
        plan_opt_test.go:319: 
            	Error Trace:	plan_opt_test.go:91
            	            				plan_opt_test.go:319
            	Error:      	Not equal: 
            	            	expected: 1
            	            	actual  : 2
            	Test:       	TestQueryCache/statschange
            	Messages:   	hits
        plan_opt_test.go:319: 
            	Error Trace:	plan_opt_test.go:92
            	            				plan_opt_test.go:319
            	Error:      	Not equal: 
            	            	expected: 2
            	            	actual  : 1
            	Test:       	TestQueryCache/statschange
            	Messages:   	misses

```

--- a/pkg/sql/create_stats.go
+++ b/pkg/sql/create_stats.go
@@ -290,6 +290,7 @@ func createStatsDefaultColumns(
 // createStatsResumer.Resume so it can be used in createStatsResumer.OnSuccess
 // (if the job is successful).
 type createStatsResumer struct {
+	tableID sqlbase.ID
 	evalCtx *extendedEvalContext
 }
 
@@ -309,6 +310,7 @@ func (r *createStatsResumer) Resume(
 		}
 	}
 
+	r.tableID = details.Table.ID
 	r.evalCtx = p.ExtendedEvalContext()
 
 	ci := sqlbase.ColTypeInfoFromColTypes([]sqlbase.ColumnType{})
@@ -419,10 +421,16 @@ func (r *createStatsResumer) OnFailOrCancel(
 func (r *createStatsResumer) OnSuccess(ctx context.Context, _ *client.Txn, job *jobs.Job) error {
 	details := job.Details().(jobspb.CreateStatsDetails)
 
+	// Invalidate the local cache synchronously; this guarantees that the next
+	// statement in the same session won't use a stale cache (whereas the gossip
+	// update is handled asynchronously).
+	r.evalCtx.ExecCfg.TableStatsCache.InvalidateTableStats(ctx, r.tableID)
+
+	// Record this statistics creation in the event log.
 	if !createStatsPostEvents.Get(&r.evalCtx.Settings.SV) {
 		return nil
 	}
-	// Record this statistics creation in the event log.
+
 	// TODO(rytaft): This creates a new transaction for the CREATE STATISTICS
 	// event. It must be different from the CREATE STATISTICS transaction,
 	// because that transaction must be read-only. In the future we may want

--- a/pkg/sql/distsqlrun/sample_aggregator.go
+++ b/pkg/sql/distsqlrun/sample_aggregator.go
@@ -270,7 +270,7 @@ func (s *sampleAggregator) writeResults(ctx context.Context) error {
 	// internal executor instead of doing this weird thing where it uses the
 	// internal executor to execute one statement at a time inside a db.Txn()
 	// closure.
-	return s.flowCtx.ClientDB.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
+	if err := s.flowCtx.ClientDB.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
 		for _, si := range s.sketches {
 			var histogram *stats.HistogramData
 			if si.spec.GenerateHistogram && len(s.sr.Get()) != 0 {
@@ -310,7 +310,6 @@ func (s *sampleAggregator) writeResults(ctx context.Context) error {
 			// Insert the new stat.
 			if err := stats.InsertNewStat(
 				ctx,
-				s.flowCtx.Gossip,
 				s.flowCtx.executor,
 				txn,
 				s.tableID,
@@ -326,7 +325,12 @@ func (s *sampleAggregator) writeResults(ctx context.Context) error {
 		}
 
 		return nil
-	})
+	}); err != nil {
+		return err
+	}
+
+	// Gossip invalidation of the stat caches for this table.
+	return stats.GossipTableStatAdded(s.flowCtx.Gossip, s.tableID)
 }
 
 // generateHistogram returns a histogram (on a given column) from a set of

--- a/pkg/sql/opt/exec/execbuilder/testdata/stats
+++ b/pkg/sql/opt/exec/execbuilder/testdata/stats
@@ -4,33 +4,28 @@
 # statistics if distsql mode is OFF.
 
 statement ok
-CREATE TABLE a (u INT, v INT, INDEX (u) STORING (v), INDEX (v) STORING (u));
-
-# Create a new table to avoid depending on the asynchronous stat cache
-# invalidation.
-statement ok
-CREATE TABLE b (u INT, v INT, INDEX (u) STORING (v), INDEX (v) STORING (u));
-INSERT INTO b VALUES (1, 1), (1, 2), (1, 3), (1, 4), (2, 4), (2, 5), (2, 6), (2, 7)
+CREATE TABLE uv (u INT, v INT, INDEX (u) STORING (v), INDEX (v) STORING (u));
+INSERT INTO uv VALUES (1, 1), (1, 2), (1, 3), (1, 4), (2, 4), (2, 5), (2, 6), (2, 7)
 
 statement ok
-CREATE STATISTICS u ON u FROM b;
-CREATE STATISTICS v ON v FROM b
+CREATE STATISTICS u ON u FROM uv;
+CREATE STATISTICS v ON v FROM uv
 
 statement ok
 set experimental_enable_zigzag_join = false
 
 # Verify we scan index v which has the more selective constraint.
 query TTTTT
-EXPLAIN (VERBOSE) SELECT * FROM b WHERE u = 1 AND v = 1
+EXPLAIN (VERBOSE) SELECT * FROM uv WHERE u = 1 AND v = 1
 ----
-scan  ·       ·          (u, v)  ·
-·     table   b@b_v_idx  ·       ·
-·     spans   /1-/2      ·       ·
-·     filter  u = 1      ·       ·
+scan  ·       ·            (u, v)  ·
+·     table   uv@uv_v_idx  ·       ·
+·     spans   /1-/2        ·       ·
+·     filter  u = 1        ·       ·
 
 # Verify that injecting different statistics changes the plan.
 statement ok
-ALTER TABLE b INJECT STATISTICS '[
+ALTER TABLE uv INJECT STATISTICS '[
   {
     "columns": ["u"],
     "created_at": "2018-01-01 1:00:00.00000+00:00",
@@ -46,17 +41,17 @@ ALTER TABLE b INJECT STATISTICS '[
 ]'
 
 query TTTTT
-EXPLAIN (VERBOSE) SELECT * FROM b WHERE u = 1 AND v = 1
+EXPLAIN (VERBOSE) SELECT * FROM uv WHERE u = 1 AND v = 1
 ----
-scan  ·       ·          (u, v)  ·
-·     table   b@b_u_idx  ·       ·
-·     spans   /1-/2      ·       ·
-·     filter  v = 1      ·       ·
+scan  ·       ·            (u, v)  ·
+·     table   uv@uv_u_idx  ·       ·
+·     spans   /1-/2        ·       ·
+·     filter  v = 1        ·       ·
 
 # Verify that injecting different statistics with null counts
 # changes the plan.
 statement ok
-ALTER TABLE b INJECT STATISTICS '[
+ALTER TABLE uv INJECT STATISTICS '[
   {
     "columns": ["u"],
     "created_at": "2018-01-01 1:00:00.00000+00:00",
@@ -74,15 +69,15 @@ ALTER TABLE b INJECT STATISTICS '[
 ]'
 
 query TTTTT
-EXPLAIN (VERBOSE) SELECT * FROM b WHERE u = 1 AND v = 1
+EXPLAIN (VERBOSE) SELECT * FROM uv WHERE u = 1 AND v = 1
 ----
-scan  ·       ·          (u, v)  ·
-·     table   b@b_u_idx  ·       ·
-·     spans   /1-/2      ·       ·
-·     filter  v = 1      ·       ·
+scan  ·       ·            (u, v)  ·
+·     table   uv@uv_u_idx  ·       ·
+·     spans   /1-/2        ·       ·
+·     filter  v = 1        ·       ·
 
 statement ok
-ALTER TABLE b INJECT STATISTICS '[
+ALTER TABLE uv INJECT STATISTICS '[
   {
     "columns": ["u"],
     "created_at": "2018-01-01 1:00:00.00000+00:00",
@@ -100,9 +95,9 @@ ALTER TABLE b INJECT STATISTICS '[
 ]'
 
 query TTTTT
-EXPLAIN (VERBOSE) SELECT * FROM b WHERE u = 1 AND v = 1
+EXPLAIN (VERBOSE) SELECT * FROM uv WHERE u = 1 AND v = 1
 ----
-scan  ·       ·          (u, v)  ·
-·     table   b@b_v_idx  ·       ·
-·     spans   /1-/2      ·       ·
-·     filter  u = 1      ·       ·
+scan  ·       ·            (u, v)  ·
+·     table   uv@uv_v_idx  ·       ·
+·     spans   /1-/2        ·       ·
+·     filter  u = 1        ·       ·

--- a/pkg/sql/stats/new_stat.go
+++ b/pkg/sql/stats/new_stat.go
@@ -26,11 +26,11 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 )
 
-// InsertNewStat inserts a new statistic in the system table and updates the
-// gossip key to notify the stat caches.
+// InsertNewStat inserts a new statistic in the system table.
+// The caller is responsible for calling GossipTableStatAdded to notify the stat
+// caches.
 func InsertNewStat(
 	ctx context.Context,
-	g *gossip.Gossip,
 	executor sqlutil.InternalExecutor,
 	txn *client.Txn,
 	tableID sqlbase.ID,
@@ -59,7 +59,7 @@ func InsertNewStat(
 		}
 	}
 
-	if _, err := executor.Exec(
+	_, err := executor.Exec(
 		ctx, "insert-statistic", txn,
 		`INSERT INTO system.table_statistics (
 					"tableID",
@@ -77,12 +77,8 @@ func InsertNewStat(
 		distinctCount,
 		nullCount,
 		histogramVal,
-	); err != nil {
-		return err
-	}
-
-	// TODO(radu): we need to clear out old stats that are superseded.
-	return GossipTableStatAdded(g, tableID)
+	)
+	return err
 }
 
 // GossipTableStatAdded causes the statistic caches for this table to be

